### DISCUSSION TITLE
Support `offset` in `findByCursor`

### DISCUSSION THREAD

**Discussion by stefansundin** [mikro-orm/mikro-orm#5347] [Ideas]
**Is your feature request related to a problem? Please describe.**

I am aware that this may be a controversial feature request, but lets have a friendly discussion about it. 😄 

I am adopting cursor based pagination, but I also want to still allow the client to jump to an arbitrary page number. In my front-end I am exposing a page selection component, so the user is able to click a single button to jump from e.g. page 1 to page 5. For performance reasons the client is passing a cursor when possible, but in this case when there isn't a cursor available it will pass an `offset` instead, but I still want a cursor to be returned by mikro-orm.

So I am basically trying to adopt cursor based pagination without changing the capabilities of my front-end.

https://github.com/mikro-orm/mikro-orm/blob/2fba5eebbc118c98a890b3083d521b52b263ba30/packages/core/src/drivers/IDatabaseDriver.ts#L156

`FindByCursorOptions` is using `Omit<..., 'offset'>` in the type, **but if you pass it anyway then it actually does work**. So I am already able to do what I want, but it is just that the types discourage me from adding this because I am afraid that it might break in a future mikro-orm version.

The only bug is that `hasPrevPage` is `false` even though there actually is a previous page. I am working around this by returning `true` to the client if they pass in an `offset`.

**Describe the solution you'd like**

Allow, but discourage, the use of `offset`. Explain the performance implications, but also when it might be useful, i.e. when skipping to an arbitrary page number without forcing the client to perform many round-trips.

Also, fix the `hasPrevPage` bug when passing an `offset`. `hasPrevPage` should be `true` when `offset > 0` or when `after` is set.

**Describe alternatives you've considered**

The alternative would be for the client to perform many round-trips with cursor pagination. This may be acceptable when jumping a few pages (e.g. page 1 to page 5), but would be expensive when jumping a large number of pages (e.g. page 1 to page 500). With the solution proposed above (which works already but the types don't officially allow it), then these round-trips are not required.

**Additional context**

This may be documented as a "mixed mode" cursor pagination maybe. I know it is counter-intuitive to allow but I think that it might be a nice-to-have feature. It should definitely be documented with care so that anyone using it understands the implications.

Thank you!

**Comment by B4nan** [comment]
Have you considered using a custom repository (or entity manager) method to do this? 

I don't think it's a good idea to add something to the ORM that would be discouraged right from the beginning. I would be actually in favor of adding runtime validation to disallow this completely, as now you depend on how the internals work.

> but if you pass it anyway then it actually does work

It does, but for cursor based pagination you need to overfetch to know if there is a next page, the limit is always `(first || last) +1`.

**Reply by stefansundin** [reply]
Thanks for your reply.

> Have you considered using a custom repository (or entity manager) method to do this?

I hadn't. When you say "custom", I'm not sure exactly how you envision the implementation?

What I did try just now is using a normal `findOne` with `offset: offset-1`, and then manually constructing an `after` cursor. Just wanted to test this out, but I find that the code becomes somewhat unwieldy.

Before I had:

```typescript
    return this.orgMembershipRepo.findByCursor<'OrgMembership', 'user'>(
      {
        org: {
          id: orgId,
        },
        user: {
          $and: [
            // some complicated filters omitted here
          ],
        },
      },
      {
        populate: ['user'],
        after,
        before,
        first,
        last,
        offset,       // <-- type issue, currently ignored
        orderBy: [
          {
            user: {
              lastName: QueryOrder.ASC,
            },
          },
          {
            user: {
              firstName: QueryOrder.ASC,
            },
          },
          { id: QueryOrder.ASC },
        ],
      },
    );
```

After:


```typescript
    if (offset && offset > 0) {
      // if an offset is passed in then first find the record to use for manually constructing a cursor from
      const orgMembership = await this.orgMembershipRepo.findOne<
        'OrgMembership',
        'user'
      >(
        {
          org: {
            id: orgId,
          },
          user: {
            $and: [
              // some complicated filters omitted here
            ],
          },
        },
        {
          populate: ['user'],
          offset: offset - 1,     // <-- using offset-1 here to get the record to use in the manually constructed cursor
          orderBy: [
            {
              user: {
                lastName: QueryOrder.ASC,
              },
            },
            {
              user: {
                firstName: QueryOrder.ASC,
              },
            },
            { id: QueryOrder.ASC },
          ],
        },
      );
      if (!orgMembership) {
        throw new Error('Expected to find a record.');
      }
      const cursorData = JSON.stringify([
        { user: { lastName: orgMembership.user.$.lastName } },
        { user: { firstName: orgMembership.user.$.firstName } },
        orgMembership.id,
      ]);
      const buff = Buffer.from(cursorData, 'utf-8');
      after = buff.toString('base64');
    }

    return this.orgMembershipRepo.findByCursor<'OrgMembership', 'user'>(
      {
        org: {
          id: orgId,
        },
        user: {
          $and: [
            // some complicated filters omitted here
          ],
        },
      },
      {
        populate: ['user'],
        after,
        before,
        first,
        last,
        orderBy: [
          {
            user: {
              lastName: QueryOrder.ASC,
            },
          },
          {
            user: {
              firstName: QueryOrder.ASC,
            },
          },
          { id: QueryOrder.ASC },
        ],
      },
    );
```

Not sure if this is what you meant?


> I don't think it's a good idea to add something to the ORM that would be discouraged right from the beginning.

What I mean is that it is a powerful option that fits a certain use case, but before someone uses it they should understand its purpose. It is still far better than the alternative, in terms of how expensive the overall operation is.

The alternative would be for the client to fetch 500 pages just to iterate on the cursor. It would throw away all of the other data. Overall it would take an immense amount of time compared to the database just skipping database records.


> for cursor based pagination you need to overfetch to know if there is a next page, the limit is always `(first || last) +1`.

I am not sure how this is relevant? The bug noted above was regarding `hasPrevPage`, not `hasNextPage`.

---

I am curious what you mean with "custom repository (or entity manager) method", so I would love to learn more about that.

I am leaning towards not exposing this `offset` capability and removing the UI that supported that. I'd rather have cursor pagination and remove the `offset` feature for now, and then potentially hacking it back in later when (and if) that becomes a pain point for users.

Thanks!
